exotica/src/KinematicTree.cpp

```diff
@@ -621,8 +621,7 @@ void KinematicTree::publishFrames()
             visualization_msgs::MarkerArray msg;
             for (int i = 0; i < Tree.size(); i++)
             {
-                // TODO: This is an incorrect check - we may have shapes that are attached to rootFrame
-                if (Tree[i].lock()->Shape && Tree[i].lock()->Parent.lock()->Id >= ModelTree.size())
+                if (Tree[i].lock()->Shape)
```

Comment:
This doesn't work because of the original reason that the added object, with transformation to base, is seen as robot link, i.e. `ClosestRobotLink->isRobotLink` is true for the added sphere.
I think the added sphere has the same properties as other robot collision objects, e.g. `lwr_arm_2_link_collision_0`.

The objects added from the scene file do not have a `ClosestRobotLink`.

The only way to solve this might be to change the way how objects are added via `addObject`. However, I am not sure how you can logically distinguish collision objects from the robot itself and other objects, that are added to a robot frame.

Reply:
I think we are adding them to a different tree - EnvironmentTree to be precise - perhaps that is a way to distinguish? We definitely want to fix the wrong statement but removing the check entirely isnt a valid fix either :-(

Reply:
The objects from the scene file are added via `loadGeometryFromStream` to moveit's internal `PlanningScene` (member `ps_` in `Scene`).
There is, however, a `addCollisionObjects` for moveit's `PlanningSceneInterface`. But I don't see this exposed somewhere in EXOTica.
Any idea how I would use something like `addCollisionObjects` on the `Scene`'s `planning_scene::PlanningScene`?

Reply:
I tried to add the object to the `PlanningScene` (ps_) via (from https://answers.ros.org/question/245995/adding-collision-object-in-moveit/):
```C++
void Scene::addObject([...])
{
    [...]

    moveit_msgs::PlanningScene msg_ps;
    ps_->getPlanningSceneMsg(msg_ps);
    moveit_msgs::CollisionObject object;
    object.id = name;
    object.operation = moveit_msgs::CollisionObject::ADD;
    shapes::ShapeMsg mesh_msg;
    shapes::constructMsgFromShape(shape.get(), mesh_msg);
    shape_msgs::Mesh mesh = boost::get<shape_msgs::Mesh>(mesh_msg);
    object.meshes.push_back(mesh);
    ps_->usePlanningSceneMsg(msg_ps);
}
```
but I am getting `failed value get using boost::get`. However even if it would work, it appears still as a workaround to me.

The visualisation, including the robot collision frames, is the true and correct respresentation of the collision scene. I think that showing a different collision scene than the one used internally is not correct.

After all, the effect you see in the image is only resulting from RViz's overlay of the robot state and the collision scene. Without overlying the robot state, you see the whole (and correct) collision scene:
![rviz_screenshot_2018_12_08-18_11_49](https://user-images.githubusercontent.com/8226248/49689139-dcf8a300-fb14-11e8-90d2-cc48c242819d.png)

Reply:
> The visualisation, including the robot collision frames, is the true and correct respresentation of the collision scene. I think that showing a different collision scene than the one used internally is not correct.

Not quite. You can specify world and robot link scaling and padding - which is not exposed back to the shapes stored in the ``KinematicElement``.

Reply:
Currently, we have scene-wide scaling and padding which is performed inside the CollisionScene implementations themselves. I agree it would be better to add that logic to the Scene itself. (E.g. cf. the ``constructFclCollisionObject`` method in the CollisionSceneFCL(Latest) as well as the initialiser variables ``WorldLinkPadding``)

Reply:
However, one argument for not moving it to Scene and changing the Shapes inside KinematicElement is that those are published for visualisation - and if they are enlarged/padded, it always seems as if _something_ is in collision even though the unpadded versions would not be.

Reply:
Would it make sense to obtain the global scale and padding from the `Scene` and add it to every `Shape` added in `addObject`? Where are these properties stored within `Scene`?

Reply:
> 
> 
> This doesn't work because of the original reason that the added object, with transformation to base, is seen as robot link, i.e. `ClosestRobotLink->isRobotLink` is true for the added sphere.
> I think the added sphere has the same properties as other robot collision objects, e.g. `lwr_arm_2_link_collision_0`.
> 
> The objects added from the scene file do not have a `ClosestRobotLink`.
> 
> The only way to solve this might be to change the way how objects are added via `addObject`. However, I am not sure how you can logically distinguish collision objects from the robot itself and other objects, that are added to a robot frame.

The robot collision shapes are treated as a special group. They are used for self-collision, which is why we are checking if the link or its parent is part of the robot tree. 

The robot tree is treated separately like that to also prevent it being modified (e.g., detach a robot arm and attach it to the world link - which would we extremely cool :) ). This keeps the robot tree in sync with the moveit robot model.

To keep the whole thing in sync but to avoid updating the whole scene every time you add an object while, for example, loading a bunch of objects from a file, the update is lazy. After you've finished adding objects run:
```
updateSceneFrames();
updateInternalFrames();
updateCollisionObjects();
```

This is missing at [this line ](https://github.com/ipab-slmc/exotica/blob/master/exotica/src/Scene.cpp#L676). I'll have to check if this was intentional or just a bug.

Reply:
I undid my changes and replaced the robot link check by `( !Tree[i].lock()->ClosestRobotLink.lock() || !Tree[i].lock()->ClosestRobotLink.lock()->isRobotLink )`.